src/Features/Core/Portable/UseCollectionInitializer/AbstractUseCollectionInitializerDiagnosticAnalyzer.cs

```diff
@@ -145,257 +156,64 @@ private void AnalyzeNode(SyntaxNodeAnalysisContext context, INamedTypeSymbol ien
-
-        internal ImmutableArray<TExpressionStatementSyntax> Analyze()
-        {
-            if (_syntaxFacts.GetObjectCreationInitializer(_objectCreationExpression) != null)
-            {
-                // Don't bother if this already has an initializer.
-                return ImmutableArray<TExpressionStatementSyntax>.Empty;
-            }
-
-            _containingStatement = _objectCreationExpression.FirstAncestorOrSelf<TStatementSyntax>();
-            if (_containingStatement == null)
-            {
-                return ImmutableArray<TExpressionStatementSyntax>.Empty;
-            }
-
-            if (!TryInitializeVariableDeclarationCase() &&
-                !TryInitializeAssignmentCase())
-            {
-                return ImmutableArray<TExpressionStatementSyntax>.Empty;
-            }
-
-            var matches = ArrayBuilder<TExpressionStatementSyntax>.GetInstance();
-            AddMatches(matches);
-            return matches.ToImmutableAndFree(); ;
-        }
-
-        private void AddMatches(ArrayBuilder<TExpressionStatementSyntax> matches)
-        {
-            var containingBlock = _containingStatement.Parent;
-            var foundStatement = false;
-
-            var seenInvocation = false;
-            var seenIndexAssignment = false;
-            
-            foreach (var child in containingBlock.ChildNodesAndTokens())
-            {
-                if (!foundStatement)
-                {
-                    if (child == _containingStatement)
-                    {
-                        foundStatement = true;
-                    }
-
-                    continue;
-                }
-
-                if (child.IsToken)
-                {
-                    return;
-                }
-
-                var statement = child.AsNode() as TExpressionStatementSyntax;
-                if (statement == null)
-                {
-                    return;
-                }
-
-                SyntaxNode instance = null;
-                if (!seenIndexAssignment)
-                {
-                    if (TryAnalyzeAddInvocation(statement, out instance))
-                    {
-                        seenInvocation = true;
-                    }
-                }
-
-                if (!seenInvocation)
-                {
-                    if (TryAnalyzeIndexAssignment(statement, out instance))
-                    {
-                        seenIndexAssignment = true;
-                    }
-                }
-
-                if (instance == null)
-                {
-                    return;
-                }
-
-                if (!ValuePatternMatches((TExpressionSyntax)instance))
-                {
-                    return;
-                }
-
-                matches.Add(statement);
-            }
-        }
 
-        private bool TryAnalyzeIndexAssignment(
-            TExpressionStatementSyntax statement,
-            out SyntaxNode instance)
+        public Task FixAllAsync(
```

Comment:
> we typically have some common helper type that is shared between them.

I could do that.  But my honest opinion is: would there really be a benefit to doing that?

Or, another way to think about it: The Analyzer is the common helper.  The CodeFixProvider exists *purely* to register the code actions.  If i could package this all up as one class, i would happily do so.  However, a problem here is that to be an analyzer, you need to subclass DiagnosticAnalyzer, and to provide code fixes, you need to subclass CodeFixProvider.  You can't subclass multiple bases, so you have to have multiple types.

> I guess it feels like analyzers and codefixes should be in separate layers.

I personally don't see it that way.  To me this is just one component that provides the entire experience around this language feature.  It's a single component that reports the issue *and* fixes it.  Due to .net issues, i need to do that through two classes.  But i'm not really feeling it's desirable to introduce a third system just to make it so that these two do not talk.

Or, in other words, if i were to do this, all i'd be doing is pulling out nearly all of the code in the Analyzer into a new type, and having the Analyzer just defer to that type.  I really don't see much benefit to that organization of code (plus it will become super ugly due to the need to have sooooooo many type parameters).

--

This is, btw, how the code was originally structured (and you can see htat in the history of this directory).  The Analyzer type was not a nested type, but was a sibling type in this namespace.  It required 8 type parameters that had to be passed along and repeated all over the place.  Definitely very annoying.  By moving it into the DiagnosticAnalyzer all of that can be shared and simplified.